external/esp_idf_port/esp32/esp32_queue_api.c

```diff
@@ -0,0 +1,434 @@
+		return NULL;
+	}
+	queues_info[mq_id].mqd_fd_recv = mq_open(name, O_RDWR | O_CREAT, 0666, &attr);
+	if (queues_info[mq_id].mqd_fd_recv == (mqd_t) ERROR) {
+		return NULL;
+	}
+	queues_info[mq_id].valid = true;
+	queues_info[mq_id].mq_item_size = item_size;
+	return &queues_info[mq_id];
+}
+
+/****************************************************************************
+ * Name: queue_delete_wrapper
+ *
+ * Description:
+ *	This function is the wrapper of mq_close to close a connection between a named
+ * message queue and the calling task.
+ *
+ * Inputs:
+ *	queue - The queue to be closed.
+ *
+ * Return:
+ *
+ ****************************************************************************/
+void IRAM_ATTR queue_delete_wrapper(void *queue)
+{
+	queue_info_t *queue_info = NULL;
+
+	if (queue) {
+		queue_info = (queue_info_t *) queue;
+		if (queue_info->mqd_fd_send != (mqd_t) ERROR) {
+			mq_close(queue_info->mqd_fd_send);
+			queue_info->mqd_fd_send = NULL;
+		}
+		if (queue_info->mqd_fd_recv != (mqd_t) ERROR) {
+			mq_close(queue_info->mqd_fd_recv);
+			queue_info->mqd_fd_recv = NULL;
+		}
+		queue_info->mq_item_size = 0;
+		queue_info->valid = false;
+	}
+	return;
+}
+
+/****************************************************************************
+ * Name: queue_send_wrapper
+ *
+ * Description:
+ *	This function is the wrapper of mq_timedsend to adds the specificied message (msg) to
+ * the message queue (mqdes).
+ *
+ * Inputs:
+ *	queue - The queue to send messages.
+ *	item - The message to be send.
+ *	block_time_tick - the time (in Ticks) to wait until a timeout is decleared
+ * Return:
+ *	Returns pdPASS if success, pdFAIL if failure.
+ *
+ ****************************************************************************/
+static bool queue_is_full(void *queue)
+{
+	queue_info_t *queue_info = (queue_info_t *) queue;
+	return false;
+}
+
+int32_t IRAM_ATTR queue_send_wrapper(void *queue, void *item, uint32_t block_time_tick)
+{
+	int32_t ret;
+	queue_info_t *queue_info = NULL;
+	struct timespec abstime;
+	uint32_t secs;
+	uint32_t msecs;
+	uint32_t nsecs;
+	uint32_t carry;
+
+	if (!queue || !item) {
+		return pdFAIL;
+	}
+	queue_info = (queue_info_t *) queue;
+	if (queue_info->mqd_fd_send != (mqd_t) ERROR) {
+		if (block_time_tick == 0xFFFFFFFF) {
+			ret = mq_send(queue_info->mqd_fd_send, (char *)item, queue_info->mq_item_size, NORMAL);
+			if (ret == ERROR) {
+				return pdFAIL;
+			}
+		}
+		else {
+			(void)clock_gettime(CLOCK_REALTIME, &abstime);
+			msecs = TICK2MSEC(block_time_tick);
+			secs = msecs / MSEC_PER_SEC;
+			nsecs = (msecs - (secs * MSEC_PER_SEC)) * NSEC_PER_MSEC;
+			abstime.tv_sec += secs;
+			abstime.tv_nsec += nsecs;
+			if (abstime.tv_nsec >= NSEC_PER_SEC) {
+				carry = abstime.tv_nsec / NSEC_PER_SEC;
+				abstime.tv_sec += carry;
+				abstime.tv_nsec -= (carry * NSEC_PER_SEC);
+			}
+			ret = mq_timedsend(queue_info->mqd_fd_send, (char *)item, queue_info->mq_item_size, NORMAL, &abstime);
```

Comment:
DRY.  
in this file,  'convert ticks to abstime' is repeated it self several times
please check 'time' module in kernel has such facility or not.
If not, at least we can abstract a static function here.

Reply:
good idea, add queue_calc_time function to refuse it,  fixed.